baseapp/abci.go

```diff
@@ -248,11 +248,29 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
 // Ref: https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-060-abci-1.0.md
 // Ref: https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_basic_concepts.md
 func (app *BaseApp) PrepareProposal(req abci.RequestPrepareProposal) (resp abci.ResponsePrepareProposal) {
-	ctx := app.getContextForTx(runTxPrepareProposal, []byte{})
 	if app.prepareProposal == nil {
 		panic("PrepareProposal method not set")
 	}
 
+	// Tendermint must never call PrepareProposal with a height of 0.
+	// Ref: https://github.com/tendermint/tendermint/blob/059798a4f5b0c9f52aa8655fa619054a0154088c/spec/core/state.md?plain=1#L37-L38
+	if req.Height < 1 {
+		panic("PrepareProposal called with invalid height")
+	}
+
+	ctx := app.prepareProposalState.ctx
+	// Here we use deliverState on the first block given that we want to be able
+	// to access any state changes made in InitChain.
+	if req.Height == 1 {
```

Comment:
nit: given that this special handling for height == 1 is common across Prepare and Process, would it be possible to create a common helper method? something like:

```
(*app) getContextForRequest(originalCtx sdk.Context, req RequestWithHeight) {
    if req.Height == 1 { return app.deliverState.ctx.CacheContext() }
    return originalCtx
}
```

Reply:
ty! Added the following method:

```go
// getContextForProposal returns the right context for PrepareProposal and
// ProcessProposal. We use deliverState on the first block to be able to access
// any state changes made in InitChain.
func (app *BaseApp) getContextForProposal(ctx sdk.Context, height int64) sdk.Context {
	if height == 1 {
		ctx, _ = app.deliverState.ctx.CacheContext()
		return ctx
	}
	return ctx
}
```